### PR TITLE
rustPlatform.importCargoLock: Adding sparse protocol

### PR DESCRIPTION
**Pull request by rayanpiro**
## Description of changes

Adding new sparse registry protocol to rust builders with minimal changes.
Properly building fetching from kellnr self hosted instance.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/build-support/rust/import-cargo-lock.nix [hunk 1]
```diff
@@ -117,9 +119,13 @@ let
       gitParts = parseGit pkg.source;
       registryIndexUrl = lib.removePrefix "registry+" pkg.source;
     in
-      if lib.hasPrefix "registry+" pkg.source && builtins.hasAttr registryIndexUrl registries then
+      if ( lib.hasPrefix "registry+" pkg.source && builtins.hasAttr registryIndexUrl registries )
+        || ( lib.hasPrefix "sparse+" pkg.source && builtins.hasAttr pkg.source registries ) then
```
Comment (winterqt) [inline pkgs/build-support/rust/import-cargo-lock.nix:123]: I'm not sure about the formatting of the parentheses here, maybe it should be this? Not really too big of a deal though, as `if (` hasn't been used anywhere in this file before. Will let others weigh in.

```suggestion
      if (lib.hasPrefix "registry+" pkg.source && builtins.hasAttr registryIndexUrl registries)
        || (lib.hasPrefix "sparse+" pkg.source && builtins.hasAttr pkg.source registries) then
```

**Comment by marcusramberg** [comment]
Commit message should be updated to follow the [commit conventions](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions).

**Review by winterqt — DISMISSED** [review]
Overall, this looks good, thank you so much! I haven't been keeping up with the progress of the sparse registry -- I had thought that it wouldn't actually change the lock file, but looks like I'm wrong there.

Could you add a small test for it? See `pkgs/build-support/rust/test/import-cargo-lock`. My first instinct would be to just add it to `basic`, but depending on how things work with lockfile generation, might be best just to add a new test.

> Commit message should be updated to follow the [commit conventions](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions).

Beat me to it, thanks! +1, please do this.

File: pkgs/build-support/rust/import-cargo-lock.nix (hunk 1 above)
Comment (rayanpiro) [reply pkgs/build-support/rust/import-cargo-lock.nix]: If in other files the common use is without spaces makes more sense, so it's modified.

**Comment by rayanpiro** [comment]
Changed commit to follow the conventions.
I'll try to review how tests are done and add a new one or modify the basic one today or tomorrow. It's not a big difference, just Cargo.lock with sparse+https:// instead of registry+https:// but I'll review it first and post if any doubt.

**Comment by rayanpiro** [comment]
Ok, some modifications were made looking for:

- Add support for download URL because could be or not the same URL than index.
- Modify less amount of code as possible.
- Add a basic sparse test as solicited.

Hope everything seems good enough now.

**Comment by winterqt** [comment]
I'll take a look in a few hours, thanks for re-pinging.

**Review by marcusramberg — APPROVED** [review]
lgtm

**Comment by rayanpiro** [comment]
Is anything left by my side?

**Review by LordMZTE — APPROVED** [review]

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**